### ISSUE TITLE
SurgeXT plays notes stuttery vs. Surge

### ISSUE THREAD

**Issue by Bleuzen** [surge-synthesizer/surge#5830]
**Bug Description:**
I tried porting my patches from Surge to Surge XT and let them run with some midi notes in Ardour for testing. I noticed that while Surge plays my midi just fine and smooth, Surge XT sometimes sounds a little different and stutters.

**Surge Version**

* Version: Surge XT 1.0.0.1edd725
* Plugin Type: VST3
* Bitness: 64-bit

**Reproduction Steps:**
1.) Load Surge and Surge XT and play the same midi notes

**Expected Behavior:**
With the same patch they should sound the same and Surge XT should play the notes as smooth as original Surge.

**Video:**
https://www.youtube.com/watch?v=BHz8VhpC7fM

**Computer Information (please complete the following information):**
 - OS: Arch Linux
 - Host: Ardour
 - Version: 6.9

**Additional Information:**
Happens with the initial patch and "Nastyfication". Haven't tested others yet, but all could be affected.

**Comment by mkruselj** [comment]
Audio example would be welcome.

Since this doesn't happen in any other DAW that I have over here, I would suggest this might be a VST3 issue in Ardour.

**Comment by Bleuzen** [comment]
> Audio example would be welcome.

[samples.zip](https://github.com/surge-synthesizer/surge/files/7949952/samples.zip)

Can already see when comparing them in Audacity that with Surge XT some notes trigger a little late:

![image](https://user-images.githubusercontent.com/12885163/151349115-7ffbc8e4-f80a-422b-bb11-fa2917c437c2.png)



> I would suggest this might be a VST3 issue in Ardour

Can't test this as the LV2 version of Surge XT currently crashes Ardour. But VST3 of Surge works fine.

**Comment by baconpaul** [comment]
This could occur if ardour sets the timestamps on vst3 midi incorrectly. In our old vst3 I don’t think we did a very good job of interleaving midi with professing based on that

I’m curious if you set your ardour block size to 64 or 32 if it still happens

**Comment by Bleuzen** [comment]
> This could occur if ardour sets the timestamps on vst3 midi incorrectly.

Ardour incorrectly setting vst3 midi timestamps seems to be the case. The block size really makes a difference here. Smaller ones make the problem less noticeable while higher ones make it more problematic.
I reported an Ardour bug now:
https://tracker.ardour.org/view.php?id=8865

**Comment by x42** [comment]
> they're overhauling their MIDI handling.

Not MIDI per se. From a Plugin's POV nothing changes. For VST3 Plugins MIDI events are still timestamped "samples" as unit. The difference in in Ardour 7 is editing in the GUI. Previously rounding errors between music-time (bar|beat) and wallclock-time (samples) caused issues.

BTW, I cannot reproduce this with other VST3 plugins. e.g. u-he Diva aligns perfectly regardless of buffer-size.

**Comment by x42** [comment]
PS. It may still be Ardour's fault or a combination of Surge/JUCE/Ardour.
The fact that it's buffersize dependent however makes me think it's more likely in the plugin.

**Comment by Bleuzen** [comment]
@x42 as shown in my screenshots above, it does not happen with all notes. Some play fine in the right time. It seems mainly between some note transitions, when one note goes directly over into another.

**Comment by Bleuzen** [comment]
I was able to test the LV2 now. Because, while Surge XT LV2 crashes Ardour on two of my machines, it works on a third one xD
Found out that Surge XT as LV2 also has this problem, so it is not only VST3, letting my think problem could be on Plugins side (I somehow trust the LV2 integration in Ardour since it is in there for a while and never saw such problems).

**Comment by baconpaul** [comment]
hmm lemme try here in logic and see. 
i am pretty sure i got it right. but who knows!

**Comment by baconpaul** [comment]
So here's the logic from surge

```
    auto midiIt = midiMessages.findNextSamplePosition(0);
    int nextMidi = -1;
    if (midiIt != midiMessages.cend())
    {
        nextMidi = (*midiIt).samplePosition;
    }

    for (int i = 0; i < buffer.getNumSamples(); i++)
    {
        if (i == nextMidi)
        {
            applyMidi(*midiIt);
            midiIt++;
            if (midiIt == midiMessages.cend())
            {
                nextMidi = -1;
            }
            else
            {
                nextMidi = (*midiIt).samplePosition;
            }
        }
```

this will land midi items on exactly the right spot *IF* their sample position starts at 0 relative to the buffer.

If the sample has an offset, then they end up all the way at the end.

Surge 1.9 crammed them at the start of the buffer so they will be 'early' but didn't try to land them at the appropriate inter-block

So I guess my question is: does ardour label its midi from block-relative samples. 

I checked in reaper and it does (with block size 512 I get the first midi location at various values between 0 and 512).

If Ardour is providing an offset, how do I get it?

@Bleuzen do you have an ability to build and test in ardour? If so I can share with you the patch I added to debug in reaper.

**Comment by x42** [comment]
MIDI even-timestamps are relative to the current plugin process cycle. The Sample positions perfectly match the audio-buffers provided to the plugin's process function.

**Comment by Bleuzen** [comment]
> @Bleuzen do you have an ability to build and test in ardour?

Yes I can do that

**Comment by baconpaul** [comment]
[x.patch.zip](https://github.com/surge-synthesizer/surge/files/7953200/x.patch.zip)

Try that patch and see what stdout you get while playing with ardour

with reaper i never get the overflow and get sample accurate application

**Comment by Bleuzen** [comment]
> [x.patch.zip](https://github.com/surge-synthesizer/surge/files/7953200/x.patch.zip)
> 
> Try that patch and see what stdout you get while playing with ardour

```
/surge/src/surge-xt/SurgeSynthProcessor.cpp:258:37: Fehler: expected »;« before »for«
  258 |     midiMessages.getFirstEventTime()
      |                                     ^
      |                                     ;
  259 | 
  260 |         for (int i = 0; i < buffer.getNumSamples(); i++)
      |         ~~~   
```

Well xD
Whatever, got it built now with an extra semicolon.
Output from Ardour:

```
NM is 0 528
Applying midi event at sample 0 of  528
NM is 880 2048
Applying midi event at sample 880 of  2048
NM is 880 2048
Midi Overflow Event 880 in  2048
NM is 240 2048
Applying midi event at sample 240 of  2048
NM is 240 2048
Midi Overflow Event 240 in  2048
NM is 1648 2048
Applying midi event at sample 1648 of  2048
NM is 1648 2048
Midi Overflow Event 1648 in  2048
NM is 1008 2048
Applying midi event at sample 1008 of  2048
NM is 1008 2048
Midi Overflow Event 1008 in  2048
NM is 368 2048
Applying midi event at sample 368 of  2048
NM is 368 2048
Midi Overflow Event 368 in  2048
NM is 1776 2048
Applying midi event at sample 1776 of  2048
NM is 1776 2048
Midi Overflow Event 1776 in  2048
NM is 1136 2048
Applying midi event at sample 1136 of  2048
NM is 1136 2048
Midi Overflow Event 1136 in  2048
NM is 496 2048
Applying midi event at sample 496 of  2048
NM is 496 2048
Midi Overflow Event 496 in  2048
NM is 1904 2048
Applying midi event at sample 1904 of  2048
NM is 1904 2048
Midi Overflow Event 1904 in  2048
NM is 1264 2048
Applying midi event at sample 1264 of  2048
NM is 1264 2048
Midi Overflow Event 1264 in  2048
NM is 624 2048
Applying midi event at sample 624 of  2048
NM is 624 2048
Midi Overflow Event 624 in  2048
NM is 2032 2048
Applying midi event at sample 2032 of  2048
NM is 2032 2048
Midi Overflow Event 2032 in  2048
NM is 1392 2048
Applying midi event at sample 1392 of  2048
NM is 1392 2048
Midi Overflow Event 1392 in  2048
NM is 752 2048
Applying midi event at sample 752 of  2048
NM is 752 2048
Midi Overflow Event 752 in  2048
NM is 112 2048
Applying midi event at sample 112 of  2048
NM is 112 2048
Midi Overflow Event 112 in  2048
NM is 1520 2048
Applying midi event at sample 1520 of  2048
NM is 1520 2048
Midi Overflow Event 1520 in  2048
NM is 880 2048
Applying midi event at sample 880 of  2048
NM is 880 2048
Midi Overflow Event 880 in  2048
NM is 240 2048
Applying midi event at sample 240 of  2048
NM is 240 2048
Midi Overflow Event 240 in  2048
NM is 1648 2048
Applying midi event at sample 1648 of  2048
NM is 1648 2048
Midi Overflow Event 1648 in  2048
NM is 1008 2048
Applying midi event at sample 1008 of  2048
NM is 1008 2048
Midi Overflow Event 1008 in  2048
NM is 0 2048
Applying midi event at sample 0 of  2048
NM is 2047 2048
Applying midi event at sample 2047 of  2048
```

**Comment by baconpaul** [comment]
huh ok.so then it's not midi sample offset. hmm.

i wonder what else could be mis-aligning buffers.

**Comment by baconpaul** [comment]
<img width="260" alt="Screen Shot 2022-01-27 at 3 02 26 PM" src="https://user-images.githubusercontent.com/13387561/151434934-2e02ad2a-8591-4127-a479-f2f8a86c737d.png">

And just to be sure I tried the vst3 in reaper using @x42 's test on my mac and saw this exact alignment (which is within the 32 sample block size alignment).

Hmmm. @selenologist you are running a recent ardour right? Can you see anything like this if you happen to be online?

**Comment by x42** [comment]
@Bleuzen Aould you try https://github.com/x42/host-synth-test.lv2 as replacement synth in the same session?

That will produce a 1 sample spike on each MIDI note event:
![image](https://user-images.githubusercontent.com/30246/151436079-b67dff41-bf4e-499b-908b-fc27c366ccec.png)

**Comment by Bleuzen** [comment]
> Could you try https://github.com/x42/host-synth-test.lv2 as replacement synth in the same session?

@x42 looks fine to me:
![image](https://user-images.githubusercontent.com/12885163/151438662-0fa0b8c4-824f-46af-b334-396daa893c75.png)
![image](https://user-images.githubusercontent.com/12885163/151438747-57c8032d-29fc-4871-ab70-ffa851a53236.png)


> Just to rule out somethin is wrong with the session itself.

That's not the case, can reproduce in any new session.

**Comment by baconpaul** [comment]
and if you just sub surge xt onto that with a sine blip (basically just take the sustain on the envelope down to zero) it misses?

why on earth would it miss in ardour for you and not in reaper for me?

@x42 have you run surge in ardour and seen this?

I guess I could give you another patch which makes surge run exactly like that blip plugin.

**Comment by x42** [comment]
> @x42 have you run surge in ardour and seen this?

Yes, I just tried, and yes it's offset:

![image](https://user-images.githubusercontent.com/30246/151441239-cea93a26-6769-47e9-86c9-933e46327302.png)

It's an older version which I had installed:  "Copy" does not work hence a screenshot: 
![image](https://user-images.githubusercontent.com/30246/151441483-117c5806-5932-402d-83cb-2c5e261bce14.png)

**Comment by baconpaul** [comment]
right so that makes some sense. the old version didn't interleave into the block properly

the new version should not do that and should be much closer.

If we apply this patch to head the synth sends a blip downstream on note on and note off and in reaper it lines up perfectly. Wanna give that a try @Bleuzen ?

[blip.patch.zip](https://github.com/surge-synthesizer/surge/files/7953735/blip.patch.zip)

**Comment by Bleuzen** [comment]
> and if you just sub surge xt onto that with a sine blip

tested with Surge and Surge XT. They both have a delay:
![image](https://user-images.githubusercontent.com/12885163/151442569-7c6efda0-66cc-49ee-8004-985ed07c7e78.png)

Only difference is (and that is why I never noticed with old Surge I guess):
Surge has somewhat constant and mostly smaller delays/offsets while with Surge XT they are more random in length:
![image](https://user-images.githubusercontent.com/12885163/151442755-b5aa7306-8fcd-48d8-a69d-9fc6d01a1522.png)
with sometimes Surge XT even being faster than Surge:
![image](https://user-images.githubusercontent.com/12885163/151442823-3d1f5e02-860c-4807-a827-cdfee06c2e9a.png)

**Comment by Bleuzen** [comment]
> If we apply this patch to head the synth sends a blip downstream on note on and note off and in reaper it lines up perfectly. Wanna give that a try @Bleuzen ?

sure, one moment...

**Comment by Bleuzen** [comment]
> If we apply this patch to head the synth sends a blip downstream on note on and note off and in reaper it lines up perfectly. Wanna give that a try @Bleuzen ?

@baconpaul the blip patched Surge XT does not have those delays:
![image](https://user-images.githubusercontent.com/12885163/151444578-13df07ba-a61c-470d-b92e-f60f85e0d981.png)

**Comment by baconpaul** [comment]
wow.
ok.
let me think about that fact.

**Comment by baconpaul** [comment]
[blipmix.patch.zip](https://github.com/surge-synthesizer/surge/files/7953942/blipmix.patch.zip)

@Bleuzen just one more if you have a chance and based on this let me think.

This one mixes the surge signal and the blip. the blip is sample accurate to the midi the surge signal should be delayed an amortized 16 samples with a min of 0 and max of 31. I am curious if that is what you see in ardour.

Thank you for these tests. I am not sure what is happening but each one is at least ruling stuff out.

**Comment by Bleuzen** [comment]
> [blipmix.patch.zip](https://github.com/surge-synthesizer/surge/files/7953942/blipmix.patch.zip)
> 
> @Bleuzen just one more if you have a chance and based on this let me think.
> 
> This one mixes the surge signal and the blip.

@baconpaul 
![image](https://user-images.githubusercontent.com/12885163/151448958-de31ffa2-a1e4-4af5-bf5d-532a4a3fad7c.png)

**Comment by baconpaul** [comment]
Right so that is what I would expect - our finest midi resolution is our block size which is 32 samples. Surge 19 just put everything at the top of the next block so should be more latent but will be differently renderer especially with large block sizes but I think xt is more correct 

so I guess now I’m unsure what the problem is.

**Comment by Bleuzen** [comment]
played a little with the notes because...

> It seems mainly between some note transitions, when one note goes directly over into another.

with notes directly one after another:
![image](https://user-images.githubusercontent.com/12885163/151451169-470ba1f3-b18f-4426-96d7-0f9d2860b8be.png)

but when shortening the notes before, for example note 1 and 2, look how this makes the offset for the next note smaller:
![image](https://user-images.githubusercontent.com/12885163/151451309-0d39f289-00ba-43f0-84d5-57ee0b87331d.png)

**Comment by baconpaul** [comment]
oh that's very interesting.
i see you also posted overflow events in the print but i'm not sure that patch was good
hold on. let me try one more patch.

**Comment by baconpaul** [comment]
oh i see one potential bug - that's very useful.
hold on

**Comment by baconpaul** [comment]
[wonder.patch.zip](https://github.com/surge-synthesizer/surge/files/7954137/wonder.patch.zip)

Does this patch improve the situation?

I think surge xt has a bug when two midi events occur with exactly the same sample which will push the items to the end of the block.

**Comment by baconpaul** [comment]
by the way if that does improve the situation, along with a few other bugs we found, we may include it in 1.0.1 release. But I'd make a nightly first.

**Comment by x42** [comment]
>  two midi events occur with exactly the same sample

Aha! Yes that would happen with the MIDI data shown the the screenshot above after the first note.
Note off (x) | Note on (y)  will have the same timestamp.

**Comment by Bleuzen** [comment]
> [wonder.patch.zip](https://github.com/surge-synthesizer/surge/files/7954137/wonder.patch.zip)
> 
> Does this patch improve the situation?

makes it much better.
Before:

![image](https://user-images.githubusercontent.com/12885163/151453100-09f9bf6a-2599-49c6-9653-9996630c26b0.png)

After / with the latest patch:

![image](https://user-images.githubusercontent.com/12885163/151453142-4855535c-ab71-4729-b996-2f90e09896f9.png)

Great, thanks :)

That said there are still little delays for each note:

![image](https://user-images.githubusercontent.com/12885163/151453246-782b8f45-f32e-4005-a4b8-4b2435bbf233.png)

Seems small, probably not audible in most cases.

**Comment by baconpaul** [comment]
Yeah that's the amortized 32 delay from the block size. No getting rid of that.

OK so cool this is a real honest to goodness bug

Let me patch this and then if you can confirm the code at head we can put together a 1.0.1

Thanks for your patience and sorry for blaming ardour :)

**Comment by x42** [comment]
> Thanks for your patience and sorry for blaming ardour :)

No worries.

Wow, from bug-report to solution in under 12 hours. Great work everyone!

**Comment by baconpaul** [comment]
Ok @Bleuzen i just merged the fix to head

if you can pull main 6e7f2e0fac33b21502495998df5affc3194b54f0 and confirm it fixes for you will will roll that up with a few other post release fixes into a 1.0.1

thanks for your patience here

**Comment by Bleuzen** [comment]
@baconpaul  works from latest git main :)

**Comment by baconpaul** [comment]
Awesome. Thank you.

We will ship that with a few there bug fixes as a 1.0.1 release this weekend sometime (maybe Monday at the outside)

Appreciate the help here.
